upgradeability_with_vtable/contracts/UpgradeabilityProxy.sol

```diff
@@ -0,0 +1,31 @@
+pragma solidity ^0.4.18;
+
+import './Proxy.sol';
+import './IRegistry.sol';
+import './UpgradeabilityStorage.sol';
+
+/**
+ * @title UpgradeabilityProxy
+ * @dev This contract represents a proxy where the implementation address to which it will delegate can be upgraded
+ */
+contract UpgradeabilityProxy is Proxy, UpgradeabilityStorage {
+
+  /**
+  * @dev Constructor function
+  */
+  function UpgradeabilityProxy(string version, bytes4[] funcs) public {
+    registry = IRegistry(msg.sender);
+    for(uint256 i = 0; i < funcs.length; i++) {
+      upgradeTo(version, funcs[i]);
+    }
+  }
+
+  /**
+  * @dev Upgrades the implementation of a given function to the requested version
+  * @param version representing the version name of the new implementation to be set
+  * @param func representing the signature of the function to be set
+  */
+  function upgradeTo(string version, bytes4 func) public {
+    _implementations[func] = registry.getVersion(version, func);
+  }
```

Comment:
I'm not sure about allowing individual functions to be upgraded to a new version, while allowing others to be kept at an older one. Even if we re-use implementations for unchanged functions, I'd rather have an atomic _upgrade to X version_ that ensures that the entire proxy is moved to a new version in a single transaction.

Reply:
I thought this as an advantage of the vtables approach. The idea was to save you money in case some functions haven’t changed. We can discuss the implementation though. 

For example, we can have an atomic *upgrade* that upgrades the non-changed functions to the new proposed version but pointing them to their current implementation address.